src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -3756,7 +3757,127 @@ public RecoverWith(Func<Exception, IGraph<SourceShape<TOut>, TMat>> partialFunct
+    [InternalApi]
+    public sealed class AsyncEnumerable<T> : GraphStage<SourceShape<T>>
+    {
+        #region internal classes
+
+        private sealed class Logic : OutGraphStageLogic
+        {
+            private readonly IAsyncEnumerator<T> _enumerator;
+            private readonly Outlet<T> _outlet;
+            private readonly Action<T> _onSuccess;
+            private readonly Action<Exception> _onFailure;
+            private readonly Action _onComplete;
+            private readonly Action<Task<bool>> _handleContinuation;
+
+            public Logic(SourceShape<T> shape, IAsyncEnumerator<T> enumerator) : base(shape)
+            {
+                _enumerator = enumerator;
+                _outlet = shape.Outlet;
+                _onSuccess = GetAsyncCallback<T>(OnSuccess);
+                _onFailure = GetAsyncCallback<Exception>(OnFailure);
+                _onComplete = GetAsyncCallback(OnComplete);
+                _handleContinuation = task =>
+                {
+                    // Since this Action is used as task continuation, we cannot safely call corresponding
+                    // OnSuccess/OnFailure/OnComplete methods directly. We need to do that via async callbacks.
+                    if (task.IsFaulted) _onFailure(task.Exception);
+                    else if (task.IsCanceled) _onFailure(new TaskCanceledException(task));
+                    else if (task.Result) _onSuccess(enumerator.Current);
+                    else _onComplete();
+                };
+
+                SetHandler(_outlet, this);
+            }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnComplete() => CompleteStage();
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnFailure(Exception exception) => FailStage(exception);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnSuccess(T element) => Push(_outlet, element);
+
+            public override void OnPull()
+            {
+                var vtask = _enumerator.MoveNextAsync();
+                if (vtask.IsCompletedSuccessfully)
+                {
+                    // When MoveNextAsync returned immediatelly, we don't need to await.
+                    // We can use fast path instead.
+                    if (vtask.Result)
+                    {
+                        // if result is true, it means we got an element. Push it downstream.
+                        Push(_outlet, _enumerator.Current);
+                    }
+                    else
+                    {
+                        // if result is false, it means enumerator was closed. Complete stage in that case.
+                        CompleteStage();
+                    }
+                }
+                else
+                {
+                    vtask.AsTask().ContinueWith(_handleContinuation);
```

Comment:
This would probably be better with a local function that does `await` but we can refactor this later.